src/tools/rbd_nbd/rbd-nbd.cc

```diff
@@ -186,10 +186,9 @@ class NBDServer
 
     if (ret < 0) {
       ctx->reply.error = htonl(-ret);
-    } else if ((ctx->command == NBD_CMD_WRITE || ctx->command == NBD_CMD_READ)
-	       && ret != static_cast<int>(ctx->request.len)) {
+    } else if ((ctx->command == NBD_CMD_WRITE) && ret != 0) {
```

Comment:
librbd can clip the output of a read if it's passed the end of the image.  I think we just remove the write return code check since it doesn't provide any value (negative error code already handled).  I would add a line for reads to pad zeroes to the end of returned ```data``` bufferlist if ```ret < ctx->request.len```.